R/base.R

```diff
@@ -168,9 +183,16 @@ class_atomic <- NULL
 #' @order 2
 class_vector <- NULL
 
+#' @export
+#' @rdname base_classes
+#' @format NULL
+#' @order 2
+class_language <- NULL
+
 # Define onload to avoid dependencies between files
 on_load_define_union_classes <- function() {
   class_numeric <<- new_union(class_integer, class_double)
   class_atomic <<- new_union(class_logical, class_numeric, class_complex, class_character, class_raw)
   class_vector <<- new_union(class_atomic, class_expression, class_list)
+  class_language <<- new_union(class_name, class_call)
```

Comment:
Should the `class_language` union type include `class_expression`?

Genuine question. `base::is.language(expression())` returns `TRUE`, but I also can see why we might not want the S7 `class_language` union to not include `class_expression`.

Reply:
IMO we shouldn't blindly copy base R here because IIRC the use of "language" is pretty inconsistent. It might be better to not bother with a union and instead force folks to be explicit. e.g. if you're trying to express "things that can appear in a parse tree generated from text", `NULL` and atomic vectors of length of 1 are also acceptable. And if you're trying to express "things that can appear in a parse tree", you could also encounter more complex inlined objects.

Reply:
> Should the `class_language` union type include `class_expression`?
> 
> Genuine question. `base::is.language(expression())` returns `TRUE`, but I also can see why we might want the S7 `class_language` union not to include `class_expression`.

Yes, important question:  Note that R's own C level   `isLanguage()  *is* quite different from base `is.language()` already (with a "big Note" in the source code in `src/main/coerce.c` .. ), but it also includes NILSXP in addition to LANGSXP.  The  `help(is.language)` example, i.e., `example(is.language)`  shows a bit the conceptual "mess":
```
> ll <- list(a = expression(x^2 - 2*x + 1), b = as.name("Jim"),
             c = as.expression(exp(1)), d = call("sin", pi))
> sapply(ll, typeof)
          a            b            c            d 
"expression"     "symbol" "expression"   "language" 
> sapply(ll, mode)
           a            b            c            d 
"expression"       "name" "expression"       "call" 
> stopifnot(sapply(ll, is.language))
```
I'm not sure we should introduce yet another definition of _"language"_  but rather the same as `is.language`, i.e., including `class_expression`.
Still, it may make sense to provide a  `class_<....>` definition for the union of `class_name` and `class_call`
.. being aware that these two are actually also not the only constituents of an expression: That additionally may contain _"atomic constants"_  such as TRUE, 3.14, or "NZ" .

Reply:
> "things that can appear in a parse tree"

What if, instead of this, we frame the definition as: "things that are not reflected unmodified by `base::eval()`".

Reply:
For computing on the language looking at the C level `eval()` would probably be better. That treats `EXPRSXP` objects as self-evaluating, unlike `base::eval()` which does something I don't care for but we are stuck with. For C `eval()` the only no-self-evaluating things are `SYMSXP`, `LANGSXP`, `PROMSXP` and `BCODESXP` , and ideally the latter two should not escape to R level.

Reply:
I was just taking the cue from `methods::getClass("language")`, but I like the justification above.